src/main/java/com/commercetools/sync/commons/utils/AssetsUpdateActionUtils.java

```diff
@@ -0,0 +1,204 @@
+                return ofNullable(matchingNewAssetDraft)
+                    .map(assetDraft -> // If asset exists, compare the two assets.
+                        buildAssetActionsBuilder.apply(oldAsset, assetDraft))
+                    .orElseGet(() -> { // If asset doesn't exists, remove asset.
+                        intermediateOldAssets.remove(oldAssetsKeyMap.get(oldAssetKey)); //TODO: EXPENSIVE
+                        return singletonList(removeAssetActionBuilder.apply(oldAssetKey));
+                    });
+            })
+            .flatMap(Collection::stream)
+            .collect(toList());
+    }
+
+
+    @Nonnull
+    private static <T> Optional<UpdateAction<T>> buildChangeAssetOrderUpdateAction(
+        @Nonnull final List<Asset> intermediateOldAssets,
+        @Nonnull final List<AssetDraft> newAssetDrafts,
+        @Nonnull final Function<List<String>, UpdateAction<T>> changeAssetOrderActionBuilder) {
+
+        final Map<String, String> oldAssetKeyToIdMap = intermediateOldAssets.stream()
+                                                                            .collect(
+                                                                                toMap(Asset::getKey, Asset::getId));
+
+        final List<String> newOrder = newAssetDrafts.stream()
+                                                    .map(AssetDraft::getKey)
+                                                    .map(oldAssetKeyToIdMap::get)
+                                                    .filter(Objects::nonNull)
+                                                    .collect(toList());
+
+        final List<String> oldOrder = intermediateOldAssets.stream()
+                                                           .map(Asset::getId)
+                                                           .collect(toList());
+
+        return buildUpdateAction(oldOrder, newOrder, () -> changeAssetOrderActionBuilder.apply(newOrder));
+    }
+
+    @Nonnull
+    private static <T> List<UpdateAction<T>> buildAddAssetUpdateActions(
+        @Nonnull final List<AssetDraft> newProductVariantAssetDrafts,
+        @Nonnull final Map<String, Asset> oldAssetsKeyMap,
+        @Nonnull final BiFunction<AssetDraft, Integer, UpdateAction<T>> addAssetActionBuilder) {
+
+
+        return IntStream.range(0, newProductVariantAssetDrafts.size())
+                        .mapToObj(assetDraftIndex ->
+                            ofNullable(newProductVariantAssetDrafts.get(assetDraftIndex))
+                                .map(assetDraft -> {
+                                    final String assetDraftKey = assetDraft.getKey();
+                                    return oldAssetsKeyMap.get(assetDraftKey) == null
+                                        ? addAssetActionBuilder.apply(assetDraft, assetDraftIndex) : null;
+                                }))
+                        .filter(Optional::isPresent)
+                        .map(Optional::get)
```

Comment:
I don't think it much of an improvement than you [previous suggestion](https://github.com/commercetools/commercetools-sync-java/pull/259#discussion_r171589896). Also, I would rather leave it like that then, so when #255 is addressed, it will also take care of it